stm32 DMA driver bindings new packaging

FRASTM opened the pull request:
With this definition of the dma bindings, a new type of stm32 DMA version is supported.
For stm32F1 and stm32L1 soc, the DMA is of type V2 without dma request : named _V2bis_
A generic stm32-dma is introduced as a basis for all the stm32 dma types. The the `#dma-cells` is specified in the .yaml for each stm32 dma version, with a specific nb of elements.
- For the V1: stm32 F2, F4, F7, H7 : the dma instance has a Fifo the dma-cells has 4 (all) elements
- For the V2: stm32 F0, F3, G0,G4,WB, L0, L4,L5  : the dma instance has no Fifo, the dma-cells has 3 elements (no 'feature')
- For the V2bis: stm32 F1, L1 : the dma instance has only one possible request and no fifo ; the dma-cells has 2 elements (no 'slot', no 'feature')
- For the dmamux, the 'feature' element is not used --> removed from #dma-cells (3 elements)

Fixes #34543

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: dma: stm32 dma driver for a new dma version

This is the dma V2bis which is particular DMA V2 instance for
stm32F1 and STM32L1 soc series. This DMA does not use the dma slot
Otherwise it is similar to version V2.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: bindings: dma for stm32 is repackaged with generic dts

This will define a new bindings for any stm32 dma feature
Depending on the soc family, the DMA is of V1 or V2 or Vbis type
And give a factored definition for feature
The difference is on the dma-cell structure.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: arm: stm32 families redefines the dma compatibility

Each stm32 dma can be of V1 or V2 or V2bis type. Each type
has a dma-cell with specific nb of element. The feature and slot
properties are not required depending on the stm32 family.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
tests: drivers: uart asynch test with overlay for each stm32 board

Based on the new bindings for the stm32 dma feature,
the overlay for enabling the dma on uart client is modified.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
tests/drivers/spi/spi_loopback  successfully  tested with nucleo_l152re, nucleo_g071rb, nucleo_l476, nucleo_f767.
(Note that stm32f767 should require a cache disable)

erwango reviewed (changes requested):
Looks good in general.

Though, I wonder if this should be considered as an API change as it will have impact on out of tree users.

FRASTM commented on drivers/dma/dma_stm32.h:
```diff
@@ -75,7 +75,7 @@ bool dma_stm32_is_fe_active(DMA_TypeDef *DMAx, uint32_t id);
 #endif
 
-#ifdef CONFIG_DMA_STM32_V2
+#if defined(CONFIG_DMA_STM32_V2) || defined(CONFIG_DMA_STM32_V2BIS)
```
yes

FRASTM commented on dts/bindings/dma/st,stm32-dma-v1.yaml:
```diff
@@ -71,32 +70,11 @@ description: |
-
     "#dma-cells":
-      const: 4
+        const: 4
```
Because the #dma-cells  does not appear in the (common) dts/bindings/dma/st,st32-dma.yaml
it has not to be in a 'property-blocklist'.

FRASTM commented on drivers/serial/uart_stm32.c:
```diff
@@ -1493,12 +1493,18 @@ static int uart_stm32_pm_control(const struct device *dev,
 #define DMA_CTLR(id, dir)						\
 	DT_INST_DMAS_CTLR_BY_NAME(id, dir)
 
+#if !DT_HAS_COMPAT_STATUS_OKAY(st_stm32_dma_v2bis)
```
I think that talking about `st_stm32_dma_v2bis` here gives more contextual information than 
`#if DT_INST_PHA_HAS_CELL(id, dma-cells , slot)`

erwango replied:
Ok. Though, instead of adding this in each driver supporting DMA, please move this to dma_stm32.h.

erwango replied:
Deifning a `STM32_DMA_SLOT` would be required so we don't have to repeat this in each driver

FRASTM commented on drivers/dma/Kconfig.stm32:
```diff
@@ -25,7 +26,8 @@ config DMA_STM32_V1
+	default y if $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_DMA_V2)) \
+			|| $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_DMA_V2BIS))
 	help
 	  Enable DMA V2 support.
```
done

FRASTM commented:
Rebase, 
Add comment in the drivers/dma/Kconfig.stm32
Change the definition of the dma-cells for the qspi instance of the disco-l475-iot1
and add one comment to the st,stm32-qspi.yaml

FRASTM commented:
rebase after merge of the #34826

erwango commented on tests/drivers/spi/spi_loopback/boards/nucleo_f411re.conf:
```diff
@@ -1,8 +0,0 @@
-CONFIG_SPI_LOOPBACK_DRV_NAME="SPI_1"
```
Why is this one removed ?

erwango replied:
Why is that removed ?

erwango reviewed (changes requested):
Few comments, but this clean up really adds value.

FRASTM commented on drivers/dma/Kconfig.stm32:
```diff
@@ -25,9 +26,14 @@ config DMA_STM32_V1
+			|| $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_DMA_V2BIS))
 	help
 	  Enable DMA V2 support.
+	  Set DT_COMPAT_ST_STM32_DMA_V2BIS when the DMA V2 instance has all
+	  the requests from the peripherals simply logically ORed to one
+	  input request, like in the stm32F1 or stm32L1 soc families.
+	  Note that a DMA of type V2bis has a same driver code as V2.
```
ok, modified

FRASTM commented on include/dt-bindings/dma/stm32_dma.h:
```diff
@@ -28,7 +28,11 @@
+#define STM32_DMA_FEATURES_FIFO_THRESHOLD(features)	0
+#endif
 
 #endif /* ZEPHYR_INCLUDE_DT_BINDINGS_DMA_STM32_DMA_H_ */
```
yes, transfer of macro is possible.

erwango commented on drivers/dma/Kconfig.stm32:
```diff
@@ -25,9 +26,16 @@ config DMA_STM32_V1
+			|| $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_DMA_V2BIS))
 	help
 	  Enable DMA V2 support.
+	  "DMA_STM32_V2 symbol controls code that is common to both st,stm32-dma
+	  and st,stm32-dma-v2bis compatible. Behavior difference between these bindings
+	  is that the st,stm32-dma-v2bis has all the requests from the peripherals
+	  logically ORed to one input request, like in the stm32F1 or stm32L1.
+	  DMA_STM32_V2BIS is enabled using compatible specific macros.
+	  Note that a DMA of type V2bis has a same driver code as V2.
```
```suggestion
	  DMA_STM32_V2 symbol controls code that is common to both "st,stm32-dma"
	  and "st,stm32-dma-v2bis" compatibles. Behavior difference between these bindings
	  is that the "st,stm32-dma-v2bis" has all the requests from the peripherals
	  logically ORed to one input request, like in the STM32F1 or STM32L1.
	  Note that a DMA of type V2bis has a same driver code as V2, and
	  V2bis specific code is controller using compatible specific macros.
```

Refering to DMA_STM32_V2BIS is misleading as this Kconfig symbol does not exist

FRASTM replied:
ok

erwango commented on CODEOWNERS:
```diff
@@ -457,7 +457,6 @@
 /include/display/                         @vanwinkeljan
 /include/dt-bindings/clock/kinetis_mcg.h  @henrikbrixandersen
 /include/dt-bindings/clock/kinetis_scg.h  @henrikbrixandersen
-/include/dt-bindings/dma/stm32_dma.h      @cybertale
```
Please also remove references to #include <dt-bindings/dma/stm32_dma.h> in various drivers

FRASTM replied:
--> drivers/spi/spi_ll_stm32.c
--> drivers/i2s/i2s_ll_stm32.c
--> drivers/sperialuart_stm32.c
--> drivers/flash/flash_stm32_qspi.c

FRASTM commented:
replace #include <dt-bindings/dma/stm32_dma.h> with #include <drivers/dma/stm32_dma.h> when needed

FRASTM commented:
rebase
add the stm32f3 disco board for the tests/drivers/spi/spi_loopback 
add the nucleo_l152re board for the tests/drivers/spi/spi_loopback

erwango reviewed (changes requested):
One comment.

Also, we need to add a specific code for existing users to detect newly wrong configuration (4 cells with compat dma-v2) and either trig an error (if there's no way to overcome the error) or generate a warning and continue if we can still compile.

FRASTM commented:
rebase
defining the DMA_SLOT for any dma type in the include/dt-bindings/dma/stm32_dma.h like the macro for the features does.

FRASTM committed:
boards: arm: disco_l475_iot1 board DTS of the dma-cell on quadspi

The stm32l47r soc has a dma of type dma-v2, so the cells have
only 3 elements,. This adpat for the quadspi periph.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
add ./tests/drivers/spi/spi_loopback/boards for nucleo_g431rb
add ./tests/drivers/uart/uart_async_api/boards for nucleo_g474re

FRASTM commented:
add one commit to move the definition of the macro (proefixed with STM32_) to the include/drivers/dma/dma_stm32.h header file (instead of include/dt-bindings/dma/stm32_dma.h, suppressed)

erwango commented on dts/bindings/dma/st,stm32-dma-v2bis.yaml:
```diff
@@ -0,0 +1,81 @@
+# in the Linux kernel declared in
+# https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/plain/Bindings/dma/st,stm32-dma.yaml
+
+# Parameter syntax of stm32 follows the dma client dts syntax
+# in the Linux kernel declared in
+# https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/plain/Bindings/dma/st,stm32-dma.yaml
```
Duplicate

FRASTM replied:
removing duplicates

erwango commented on dts/bindings/dma/st,stm32-dma.yaml:
```diff
@@ -0,0 +1,38 @@
+    dma-offset:
+      type: int
+      required: false
+      description: >
```
New recommendation is to use:
```suggestion
      description: |
```

FRASTM replied:
changed

erwango replied:
Not resolved

erwango commented on dts/bindings/dma/st,stm32-dmamux.yaml:
```diff
@@ -55,10 +54,11 @@ description: |
-         dmas = <&dmamux1 11 7 0x20440 0
-         &dmamux1 1 6 0x20480 0>;
+         dmas = <&dmamux1 11 7 0x20440
+         &dmamux1 1 6 0x20480>;
```
Xould be nice to fix indentation, while we're at it

FRASTM replied:
done

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -858,37 +858,30 @@ static void spi_stm32_irq_config_func_##id(const struct device *dev)		\
-		DT_INST_DMAS_CTLR_BY_NAME(id, dir)
-
 #define SPI_DMA_CHANNEL_INIT(index, dir, dir_cap, src_dev, dest_dev)	\
-	.dma_dev = DEVICE_DT_GET(DMA_CTLR(index, dir)),			\
```
For consistency, either do the same change in uart driver in this commit as well.
Either move this to another commit

FRASTM replied:
in this commit ab764e3335214bfc8ef4a036b851d7f03e96521f
```
-	.dma_dev = DEVICE_DT_GET(DMA_CTLR(index, dir)),			\
+	.dma_dev = DEVICE_DT_GET(STM32_DMA_CTLR(index, dir)),			\
```
 for _both_ drivers drivers/spi/for spi_ll_stm32.c  and drivers/serial/uart_stm32.c

erwango commented on soc/arm/st_stm32/stm32f7/soc.c:
```diff
@@ -32,11 +32,13 @@ static int st_stm32f7_init(const struct device *arg)
 
 	key = irq_lock();
 
+#ifndef CONFIG_NOCACHE_MEMORY
```
This can't be done in current PR

FRASTM replied:
ok

erwango commented on dts/arm/st/l1/stm32l1.dtsi:
```diff
@@ -419,6 +419,16 @@
 			label = "EEPROM_0";
 		};
+
+		dma1: dma@40026000 {
```
Please do this in a dedicated commit/PR

FRASTM replied:
is it correct if `dma1: dma@40026000 `  is set in the overlay of the **nucleo_152re** for the _test/drivers/spi/spi_loopback_ ?

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -1502,7 +1502,7 @@ static int uart_stm32_pm_control(const struct device *dev,
 	.dma_channel = DT_INST_DMAS_CELL_BY_NAME(index, dir, channel),	\
 	.dma_cfg = {							\
-		.dma_slot = DT_INST_DMAS_CELL_BY_NAME(index, dir, slot),\
+		.dma_slot = DMA_SLOT(index, dir, slot),\
```
DMA_SLOT is not defined

FRASTM replied:
changed

FRASTM committed:
drivers: dma of the stm32 refactored macro for the dma-cells

The macro to set the element of the dma-cells for each peripheral
are defined in the dma_stm32 header file
and used in the periph driver (as dma client)

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
including changes from comments
spi tests for nucleo_l15re is in a separated PR

erwango commented on drivers/serial/Kconfig.stm32:
```diff
@@ -5,6 +5,7 @@
 
 DT_COMPAT_ST_STM32_DMA_V1 := st,stm32-dma-v1
 DT_COMPAT_ST_STM32_DMA_V2 := st,stm32-dma-v2
+DT_COMPAT_ST_STM32_DMA_V2BIS := st,stm32-dma-v2bis
```
Please move this in a dedicated commit

FRASTM replied:
removing all the _tests_ stuff

FRASTM replied:
In case there is a element missing in the dma-cells, the building will fail with something like:
```
nucleo_f746zg.dts.pre.tmp:2550.2-2551.22: Warning (dmas_property): /soc/spi@40013000:dmas: cell 5 is not a phandle reference
nucleo_f746zg.dts.pre.tmp:295.22-304.5: Warning (dmas_property): /soc/spi@40013000: Missing property '#dma-cells' in node /clocks/pll or bad phandle (referred from dmas[5])
  also defined at nucleo_f746zg.dts.pre.tmp:2411.20-23
  also defined at nucleo_f746zg.dts.pre.tmp:2504.7-2508.3
  also defined at nucleo_f746zg.dts.pre.tmp:2549.7-2553.3
devicetree error: <Node /clocks/pll in 'nucleo_f746zg.dts.pre.tmp'> lacks #dma-cells

```

erwango replied:
ok, then we'll need to send a mail to users & dev list to inform of this change.

erwango replied:
What about the case when there is one too much ?

FRASTM replied:
in case there is an extra parameter, the error comes in the building as well:
```
stm32f3_disco.dts.pre.tmp:1515.2-1516.22: Warning (dmas_property): /soc/spi@40013000:dmas: cell 3 is not a phandle reference
stm32f3_disco.dts.pre.tmp:1515.2-1516.22: Warning (dmas_property): /soc/spi@40013000:dmas: Could not get phandle node for (cell 3)
devicetree error: dma-names property in /soc/spi@40013000 in stm32f3_disco.dts.pre.tmp has 2 strings, expected 4 strings
```

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -1490,38 +1490,32 @@ static int uart_stm32_pm_control(const struct device *dev,
+	.dma_dev = DEVICE_DT_GET(STM32_DMA_CTLR(index, dir)),			\
 	.dma_channel = DT_INST_DMAS_CELL_BY_NAME(index, dir, channel),	\
 	.dma_cfg = {							\
 		.dma_slot = DT_INST_DMAS_CELL_BY_NAME(index, dir, slot),\
```
`STM32_DMA_SLOT`

FRASTM replied:
shouldn't the .dma_dev be made by the STM32_DMA_CTLR ?

erwango replied:
Sure

erwango commented on CODEOWNERS:
```diff
@@ -461,7 +461,6 @@
 /include/display/                         @vanwinkeljan
 /include/dt-bindings/clock/kinetis_mcg.h  @henrikbrixandersen
 /include/dt-bindings/clock/kinetis_scg.h  @henrikbrixandersen
-/include/dt-bindings/dma/stm32_dma.h      @cybertale
```
This change should be moved to another appropriate commit

FRASTM replied:
moved to commit a6a6897b79e1640192697e45197a834be64d99ff

erwango commented on include/dt-bindings/dma/stm32_dma.h:
```diff
@@ -1,34 +0,0 @@
-/*
```
This change should be moved to another commit

FRASTM replied:
moved to commit a6a6897

erwango commented on tests/drivers/spi/spi_loopback/boards/nucleo_f746zg.conf:
```diff
@@ -1,2 +1,10 @@
 CONFIG_SPI_LOOPBACK_DRV_NAME="SPI_1"
-CONFIG_SPI_STM32_INTERRUPT=y
```
I'm not sure we need to convert all tests to use DMA, as we won't test IRQ mode anymore in the end

FRASTM replied:
Ok to  modify only the existing test cases

erwango reviewed (changes requested):
Some clean up required.
Also we miss a test to generate a warning that would detect wrong configurations (excessive number of cells), to help current users identifying the change and updating their out of tree configurations.

FRASTM committed:
drivers: dma: stm32 dma driver using DT compat macro

Controlling the DMA offset  for the request, relies on the
dma version different from V1

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -17,7 +17,6 @@
 #include <drivers/clock_control/stm32_clock_control.h>
 #include <drivers/clock_control.h>
 #include <drivers/flash.h>
-#include <dt-bindings/dma/stm32_dma.h>
```
now in the commit 611b8aa1f9e127ef6a2544555ce4fa2487a62211

FRASTM replied:
now in the commit ce0731509db4f7aef958ea27e416ae822d5e0dda

FRASTM commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -9,7 +9,7 @@
 #include <string.h>
 #include <drivers/dma.h>
 #include <drivers/i2s.h>
-#include <dt-bindings/dma/stm32_dma.h>
```
now in the commit 696124e008b64766dc272cbb9faffc710296aa39

FRASTM commented on drivers/dma/Kconfig.stm32:
```diff
@@ -25,7 +26,8 @@ config DMA_STM32_V1
+	default y if $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_DMA_V2)) \
+			|| $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_DMA_V2BIS))
 	help
 	  Enable DMA V2 support.
```
do you mean commenting the commit message or the "help" of the drivers/dma/Kconfig.stm32 ?

erwango replied:
I mean adding comment in help section to help understanding the use of DMA_STM32_V2 that match both compatibles, and use of compatibles based macros for differences between V2 and V2BIS

FRASTM replied:
this is now in the commit 78bca5627ad0943969768648e1c01dbc1e745642

FRASTM commented:
one more commit for drivers/flash/flash_stm32_qspi.c 
one more commit for drivers/i2s/i2s_ll_stm32.c

FRASTM committed:
include: dt-bindings: remove dma header for stm32 soc

All the macro for dma-cells are now in the
include/drivers/dma/dma_stm32.h header file.
So the include/dt-bindings/dma/stm32_dma.h is no more
useful and removed from #include.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
rebase (fix conflict with CODEOWNERS)

mbolivar-nordic reviewed (approved):
Nice work on the bindings documentation

FRASTM commented:
rebased on cc0abee3db739413f2498d465386ab5a8bf7c1eb

FRASTM committed:
tests: drivers: spi loopback test with overlay for each stm32 board

Based on the new bindings for the stm32 dma feature,
the overlay for enabling the dmamux on spi client is modified.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

ABOSTM reviewed (approved):
LGTM

erwango reviewed (changes requested):
Nack'ed as we wait for the CI fix

erwango commented:
> Nack'ed as we wait for the CI fix

EDIT: Set as DNM, waiting for the CI fix

FRASTM commented:
rebase and fix CI for nucleo_l476rg ( dma is of type V2)

FRASTM committed:
dts: stm32l5 configure the dma V2 with new bindings

The dma V2 instances have only 3 parameters in thei dma-cells.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
tests: spi: spi loopback test running on the stm32l5xx boards

add support of the dma (V2) for the stm32l5 for spi transfer.
on all available target boards

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
tests: uart: uart async test running on the stm32l5xx boards

add support of the dma  (V2bis) for the stm32l5 for uart transfers.
on all available target boards

Signed-off-by: Francois Ramu <francois.ramu@st.com>
